### PR TITLE
feat: set teams that automatically grant permissions

### LINKED ISSUE #496: automatically grant leads team `team_lead` permission
**Please give a one-sentence summary of the feature you would like to see.**

automatically grant leads team `team_lead` permission

**Please give as many details as possible about the requirements for the feature.**

E.g. "every user should be able to have an arbitrary number of nicknames, and users should be able to give each other nicknames."

**Please list any additional context.**

For instance, if this was requested by a Launch Pad president, if it has been discussed somewhere else like Slack, etc.
E.g. "nicknames are required for the new website."

### PR DESCRIPTION
**Pull request by bobheadxi**
# Pull Request

## Description

**Give a brief description of your changes:**

Adds ability for all members in certain teams to get Rocket permissions. Need this to automatically grant everyone in @ubclaunchpad/leads the ability to create teams.

We'll need to add the following to our config:

```
GITHUB_ADMIN_TEAM_NAME='exec'
GITHUB_LEADS_TEAM_NAME='leads'
```

I know we talked about possibly dropping refresh (#490), but we still need a way to perform the initial sync - I imagine refresh might stay around purely for use cases like this.

There is also a workaround for #497 that I think should work in https://github.com/ubclaunchpad/rocket2/pull/504/commits/35e1209682d9906e7ccf7bf0ee64c601de0084ee

## Testing

**If testing this change requires extra setup, please document it here:**

## Ticket(s)

Closes #496 

(Create a copy of that line for each Github Issue affected,
and replace "Affects" with "Closes" if merging this will close the relevant ticket.)

### CODE REVIEW

File: app/controller/command/commands/team.py [hunk 1]
```diff
@@ -385,8 +386,9 @@ def add_helper(self, param_list, user_id) -> ResponseTuple:
         """
         try:
             command_user = self.facade.retrieve(User, user_id)
+            command_team = param_list['team_name']
             teams = self.facade.query(Team, [('github_team_name',
-                                              param_list['team_name'])])
+                                              command_team)])
             if len(teams) != 1:
                 return self.lookup_error, 200
             team = teams[0]
```
Comment (cheukyin699) [inline app/controller/command/commands/team.py:394]: Check `db.utils.get_team_by_name()` that does the same thing with error handling.

File: app/controller/command/commands/team.py [hunk 2]
```diff
@@ -424,8 +439,9 @@ def remove_helper(self, param_list, user_id) -> ResponseTuple:
...
```
Comment (cheukyin699) [inline app/controller/command/commands/team.py:447]: Same as above.

File: app/controller/command/commands/team.py [hunk 3]
```diff
@@ -442,7 +458,17 @@ def remove_helper(self, param_list, user_id) -> ResponseTuple:
...
+            msg = "Removed User from " + command_team
+
+            # If this team is a team with special permissions, demote the user
+            if command_team == self.config.github_team_admin\
+                    or command_team == self.config.github_team_leads:
+                logging.info(f"Demoting {command_user} to member")
+                user.permissions_level = Permissions.member
+                self.facade.store(user)
+                msg += " and demoted user"
+
```
Comment (cheukyin699) [inline app/controller/command/commands/team.py:471]: User may be both admin and team lead - if they are removed from admin rank, they should go to team lead. If they are removed from team leads, they should remain an admin.

File: config/__init__.py [hunk 4]
```diff
@@ -39,6 +41,8 @@ class Config:
     OPTIONALS = {
         'AWS_LOCAL': 'False',
         'GITHUB_DEFAULT_TEAM_NAME': 'all',
+        'GITHUB_ADMIN_TEAM_NAME': '',
+        'GITHUB_LEADS_TEAM_NAME': '',
```
Comment (cheukyin699) [inline config/__init__.py:45]:
```suggestion
        'GITHUB_ADMIN_TEAM_NAME': 'exec',
        'GITHUB_LEADS_TEAM_NAME': 'leads',

```

File: db/utils.py [hunk 5]
```diff
@@ -35,6 +35,25 @@ def get_team_by_name(dbf: DBFacade, gh_team_name: str) -> Team:
...
+    for github_id in team.members:
+        users = dbf.query(User, [('github_user_id', github_id)])
+        if len(users) != 1:
+            logging.warn(f"None/multiple users for GitHub ID {github_id}")
+
+        # For now, naiively iterate over all users, due to
+        # https://github.com/ubclaunchpad/rocket2/issues/493
+        for user in users:
+            users.append(user)
+    return users
```
Comment (cheukyin699) [inline db/utils.py:54]: How is this different from:

```suggestion
    return get_users_by_ghid(dbf, list(team.members))
```

File: docs/Config.rst [hunk 6]
```diff
@@ -72,6 +72,29 @@ GITHUB_ORG_NAME
...
+------------------------
+
+The name of the GitHub team in your organization that all users should
+be added to. Optional, defaults to ``all``.
+
+GITHUB_ADMIN_TEAM_NAME
+----------------------
+
+The name of the GitHub team in your organization that should be automatically
+promoted to Rocket administrators. Optional.
```
Comment (cheukyin699) [inline docs/Config.rst:85]:
```suggestion
promoted to Rocket administrators. Optional, defaults to ``exec``.
```

File: docs/Config.rst [hunk 7]
```diff
@@ -72,6 +72,29 @@ GITHUB_ORG_NAME
...
+
+Note that this does not mean all Rocket administrators will be added to this
+team.
+
+GITHUB_LEADS_TEAM_NAME
+----------------------
+
+The name of the GitHub team in your organization that should be automatically
+promoted to Rocket team leads. Optional.
```
Comment (cheukyin699) [inline docs/Config.rst:94]:
```suggestion
promoted to Rocket team leads. Optional, defaults to ``leads``.
```

File: app/controller/command/commands/team.py [hunk 8]
```diff
@@ -671,6 +700,41 @@ def refresh_all_team(self):
...
+                t_id = str(self.gh.org_create_team(t['name']))
+                logging.info(f'team {t["name"]} created')
+                self.facade.store(Team(t_id, t['name'], t['name']))
+
+            if team is not None:
+                team_members = get_team_members(team)
+                for user in team_members:
+                    user.permissions_level = t['permission']
+                    self.facade.store(user)
+
```
Comment (cheukyin699) [inline app/controller/command/commands/team.py:737]: Correct me if I'm wrong, but since the admin goes first, it would first promote a user to admin status, then demote the user to lead status?

File: app/controller/command/commands/team.py [hunk 9]
```diff
@@ -682,5 +746,23 @@ def refresh_all_drive_permissions(self):
...
+        # Workaround for https://github.com/ubclaunchpad/rocket2/issues/497:
+        # We sort the teams such that special-permissions teams are sync'd
+        # last, so that inherited permissions are not overwritten in child
+        # folders.
+        #
+        # TODO: If a proper fix is implemented, remove this and related code
+        for t in [leads_team, admin_team]:
+            if t is not None:
+                sync_team_email_perms(self.gcp, self.facade, t)
```
Comment (cheukyin699) [inline app/controller/command/commands/team.py:768]: Add this to function docstring please.

**Review by chuck-sys — CHANGES_REQUESTED** [review]

File: app/controller/command/commands/team.py (hunk 9 above)
Comment (bobheadxi) [reply app/controller/command/commands/team.py:768]: The workaround? I think this is better here, since it doesn't pertain to the entire function - the end result is the same ("Refresh Google Drive permissions for all teams"), with a minor implementation detail in that it has priorities

Comment (bobheadxi) [reply app/controller/command/commands/team.py:768]: This also makes it clearer what needs to be removed further down the line

File: config/__init__.py (hunk 4 above)
Comment (bobheadxi) [reply config/__init__.py:45]: I think I'd rather leave this without defaults, and have it in our configuration - I still have some distant hope of a generalized Rocket 2 for many orgs 🙃

Comment (bobheadxi) [reply config/__init__.py:45]: It is in the `sample-env` now for what it's worth

File: app/controller/command/commands/team.py (hunk 1 above)
Comment (bobheadxi) [reply app/controller/command/commands/team.py:394]: 285f80f41aeec17c3a35c2002d18c87b323d73cb

File: app/controller/command/commands/team.py (hunk 2 above)
Comment (bobheadxi) [reply app/controller/command/commands/team.py]: 285f80f41aeec17c3a35c2002d18c87b323d73cb

File: db/utils.py (hunk 5 above)
Comment (bobheadxi) [reply db/utils.py:54]: Oh hey, it is not any different!

File: docs/Config.rst (hunk 6 above)
Comment (bobheadxi) [reply docs/Config.rst:85]: convo in https://github.com/ubclaunchpad/rocket2/pull/504/files#r496301339

File: docs/Config.rst (hunk 7 above)
Comment (bobheadxi) [reply docs/Config.rst:94]: convo in https://github.com/ubclaunchpad/rocket2/pull/504/files#r496301339

File: app/controller/command/commands/team.py (hunk 8 above)
Comment (bobheadxi) [reply app/controller/command/commands/team.py:717]: 3d9fcc7528202b410ad521b04789defe3ef80e4a

File: app/controller/command/commands/team.py (hunk 3 above)
Comment (bobheadxi) [reply app/controller/command/commands/team.py]: Good catch, thanks! ea1f30c721f5010eb496b92ff72c5209a900cd72

File: app/controller/command/commands/team.py [hunk 10]
```diff
@@ -385,19 +387,29 @@ def add_helper(self, param_list, user_id) -> ResponseTuple:
...
+                promoted_level = Permissions.admin
+            elif command_team == self.config.github_team_leads:
+                promoted_level = Permissions.team_lead
+
+            msg = "Added User to " + command_team
+            if promoted_level != Permissions.member:
+                logging.info(f"Promoting {command_user} to {promoted_level}")
+                user.permissions_level = promoted_level
+                self.facade.store(user)
+                msg += f" and promoted user to {promoted_level}"
```
Comment (cheukyin699) [inline app/controller/command/commands/team.py:412]: Still runs into the issue.

Say someone needs to be admin and is also a team lead for some team. An existing admin adds them into the admin team because they are admin. That person gets promoted to admin privileges. Now the existing admin adds them into the team leads team because they are also team lead. That person gets demoted to lead privileges.

File: app/controller/command/commands/team.py [hunk 11]
```diff
@@ -442,7 +451,34 @@ def remove_helper(self, param_list, user_id) -> ResponseTuple:
...
             self.facade.store(team)
-            msg = "Removed User from " + param_list['team_name']
+
+            msg = "Removed User from " + command_team
+
+            # If this team is a team with special permissions, demote the user
+            demoted_level = None
+            if command_team == self.config.github_team_leads:
+                # Leads can only be demoted to member
+                demoted_level = Permissions.member
```
Comment (cheukyin699) [inline app/controller/command/commands/team.py:461]: Similar to above. If the admin is also a team lead, they will be demoted to member if they are removed from the team leads team.

**Review by chuck-sys — CHANGES_REQUESTED** [review]

File: app/controller/command/commands/team.py (hunk 10 above)
Comment (bobheadxi) [reply app/controller/command/commands/team.py]: 838444538a1b7d3160929b901e04851eea3326c9 and 5e7013d25f7e862ea910399badf160efe3bb9751

File: app/controller/command/commands/team.py (hunk 11 above)
Comment (bobheadxi) [reply app/controller/command/commands/team.py]: fe49d42f937104ba2b244ac6a67bb58bc69929b0

**Comment by bobheadxi** [comment]
Tests for edge cases: e3a89c521d6077509c5221f34bd08618d0d8aeb9

File: app/controller/command/commands/team.py [hunk 12]
```diff
@@ -671,6 +707,42 @@ def refresh_all_team(self):
...
+            except LookupError:
+                t_id = str(self.gh.org_create_team(t['name']))
+                logging.info(f'team {t["name"]} created')
+                self.facade.store(Team(t_id, t['name'], t['name']))
+
+            if team is not None:
+                team_members = get_team_members(team)
+                for user in team_members:
+                    user.permissions_level = t['permission']
+                    self.facade.store(user)
```
Comment (cheukyin699) [inline app/controller/command/commands/team.py:744]: Slight loophole (technically possible, though probably won't be if nobody uses rocket 2):

If the organization has admins appointed already, but do not have an admin team specifically. And if they have a leads team, and they do an update, then all admins who are also team leads will be demoted to team lead status, and because there is no admin team they won't get their permissions back.

**Review by chuck-sys — CHANGES_REQUESTED** [review]

File: app/controller/command/commands/team.py (hunk 12 above)
Comment (bobheadxi) [reply app/controller/command/commands/team.py:744]: arghhhhh

Comment (bobheadxi) [reply app/controller/command/commands/team.py]: Wow I even wrote it in the docs!

```py
    def refresh_all_rocket_permissions(self):
        """
        Refresh Rocket permissions for members in teams like
        GITHUB_ADMIN_TEAM_NAME and GITHUB_LEADS_TEAM_NAME.

        It only ever promotes users, and does not demote users.
        """
```

Fix: d0c8b2344858825824b7f510ec33ebef29a1af25

File: app/controller/command/commands/team.py [hunk 13]
```diff
@@ -671,6 +707,43 @@ def refresh_all_team(self):
...
+                t_id = str(self.gh.org_create_team(t['name']))
+                logging.info(f'team {t["name"]} created')
+                self.facade.store(Team(t_id, t['name'], t['name']))
+
+            if team is not None:
+                team_members = get_team_members(team)
+                for user in team_members:
+                    if user.permissions_level < t['permission']:
+                        user.permissions_level = t['permission']
+                        self.facade.store(user)
```
Comment (cheukyin699) [inline app/controller/command/commands/team.py:745]: mmmm yes that sextuple indent

**Review by chuck-sys — APPROVED** [review]
> do not lie

no papa
